validate/src/channel.rs

```diff
@@ -0,0 +1,197 @@
+pub const CHANNEL_RATE_LIMIT_PER_USER_MAX: u64 = 21600;
+
+/// Maximum length of a channel's topic.
+pub const CHANNEL_TOPIC_LENGTH_MAX: usize = 1024;
+
+/// Returned when the channel can not be updated as configured.
+#[derive(Debug)]
+pub struct ChannelValidationError {
+    /// Type of error that occurred.
+    kind: ChannelValidationErrorType,
+}
+
+impl ChannelValidationError {
+    /// Immutable reference to the type of error that occurred.
+    #[must_use = "retrieving the type has no effect if left unused"]
+    pub const fn kind(&self) -> &ChannelValidationErrorType {
+        &self.kind
+    }
+
+    /// Consume the error, returning the source error if there is any.
+    #[allow(clippy::unused_self)]
+    #[must_use = "consuming the error and retrieving the source has no effect if left unused"]
+    pub fn into_source(self) -> Option<Box<dyn Error + Send + Sync>> {
+        None
+    }
+
+    /// Consume the error, returning the owned error type and the source error.
+    #[must_use = "consuming the error into its parts has no effect if left unused"]
+    pub fn into_parts(
+        self,
+    ) -> (
+        ChannelValidationErrorType,
+        Option<Box<dyn Error + Send + Sync>>,
+    ) {
+        (self.kind, None)
+    }
+}
+
+impl Display for ChannelValidationError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match &self.kind {
+            ChannelValidationErrorType::NameInvalid => {
+                f.write_str("the length of the name is invalid")
+            }
+            ChannelValidationErrorType::RateLimitPerUserInvalid { .. } => {
+                f.write_str("the rate limit per user is invalid")
+            }
+            ChannelValidationErrorType::TopicInvalid => f.write_str("the topic is invalid"),
+            &ChannelValidationErrorType::TypeInvalid { kind } => {
+                Display::fmt(kind.name(), f)?;
+
+                f.write_str(" is not a thread")
+            }
+        }
+    }
+}
+
+impl Error for ChannelValidationError {}
+
+/// Type of [`ChannelValidationError`] that occurred.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum ChannelValidationErrorType {
+    /// The length of the name is either fewer than 1 UTF-16 characters or
+    /// more than 100 UTF-16 characters.
+    NameInvalid,
+    /// The seconds of the rate limit per user is more than 21600.
+    RateLimitPerUserInvalid {
+        /// Provided ratelimit is invalid.
+        rate_limit_per_user: u64,
+    },
+    /// The length of the topic is more than 1024 UTF-16 characters.
+    TopicInvalid,
+    /// Provided type was not a thread.
+    TypeInvalid {
+        /// Provided type.
+        kind: ChannelType,
+    },
+}
+
+/// Ensure a channel is a thread.
+///
+/// # Errors
+///
+/// Returns an error of type [`ChannelValidationErrorType::TypeInvalid`] if the
+/// channel is not a thread.
+pub const fn is_thread(kind: ChannelType) -> Result<(), ChannelValidationError> {
```

Comment:
i don't know how to better write this especially since the function it's called in returns the error type, so we would just have to leave around the error variant anyway. do you have suggestions?

Reply:
I think this is fine as is